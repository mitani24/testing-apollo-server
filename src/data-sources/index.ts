import { Config } from "apollo-server-express";
import { TrackAPI } from "./track-api";

export type ContextType = {
  dataSources: {
    trackAPI: TrackAPI;
  };
};

export const dataSources: Config["dataSources"] = () => {
  return {
    trackAPI: new TrackAPI(),
  };
};
