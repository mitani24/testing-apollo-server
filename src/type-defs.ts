import fs from "fs";
import path from "path";
import { gql } from "apollo-server";

const schema = fs
  .readFileSync(path.join(__dirname, "../schema.graphql"))
  .toString();

export const typeDefs = gql`
  ${schema}
`;
