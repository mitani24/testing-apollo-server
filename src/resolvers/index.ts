import { ContextType } from "../data-sources";
import { Resolvers } from "../lib/generated/graphql-resolver-types";

export const resolvers: Resolvers<ContextType> = {
  Author: {
    id: ({ id }) => id,
    name: ({ name }) => name,
    photo: ({ photo }) => photo ?? null,
  },
  Mutation: {
    async updateTrack(_, { input }, { dataSources }) {
      const track = await dataSources.trackAPI.updateTrack({
        path: {
          track_id: input.id,
        },
        body: {
          title: input.title ?? undefined,
          description: input.description ?? undefined,
          author_id: input.authorId ?? undefined,
          thumbnail: input.thumbnail ?? undefined,
          length: input.length ?? undefined,
          number_of_views: input.numberOfViews ?? undefined,
        },
      });
      return { track };
    },
  },
  Query: {
    tracks(_, { authorId }, { dataSources }) {
      return dataSources.trackAPI.getTracks({
        query: {
          author_id: authorId ?? undefined,
        },
      });
    },
    track(_, { id }, { dataSources }) {
      return dataSources.trackAPI.getTrack({
        path: {
          track_id: id,
        },
      });
    },
  },
  Track: {
    id: ({ id }) => id,
    title: ({ title }) => title,
    author: ({ author_id }, _, { dataSources }) => {
      return dataSources.trackAPI.getAuthor({
        path: {
          author_id,
        },
      });
    },
    thumbnail: ({ thumbnail }) => thumbnail ?? null,
    length: ({ length }) => length ?? null,
    description: ({ description }) => description ?? null,
    numberOfViews: ({ number_of_views }) => number_of_views ?? null,
  },
  UpdateTrackPayload: {
    track: ({ track }) => track,
  },
};
