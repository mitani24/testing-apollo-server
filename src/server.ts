import { ApolloServer } from "apollo-server";
import { typeDefs } from "./type-defs";
import { dataSources } from "./data-sources";
import { resolvers } from "./resolvers";

export const server = new ApolloServer({
  typeDefs,
  dataSources,
  resolvers,
});
