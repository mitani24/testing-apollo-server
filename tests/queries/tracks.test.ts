import { HTTPCache } from "apollo-datasource-rest";
import { rest } from "msw";
import { setupServer } from "msw/node";
import { server } from "../../src/server";
import { Tracks } from "../lib/generated/graphql-documents";

const mockServer = setupServer(
  rest.all("*", async (req, res, ctx) => {
    const redirectUrl = new URL(req.url);
    redirectUrl.protocol = "http";
    redirectUrl.host = "localhost";
    redirectUrl.port = "4010";
    const redirectReq = { ...req, url: redirectUrl };

    const data = await ctx.fetch(redirectReq);
    const json = await data.json();

    const headers: Record<string, string> = {};
    for (const key of data.headers.keys()) {
      headers[key] = data.headers.get(key);
    }

    return res(ctx.status(data.status), ctx.set(headers), ctx.json(json));
  })
);

beforeAll(() => mockServer.listen({ onUnhandledRequest: "error" }));
afterEach(() => mockServer.resetHandlers());
afterAll(() => mockServer.close());

describe("tracks", () => {
  it("トラック一覧を取得できる", async () => {
    const res = await server.executeOperation({ query: Tracks });
    expect(res).toMatchSnapshot();
  });

  it("author_id でフィルタできる", async () => {
    const fetch = jest.spyOn(HTTPCache.prototype, "fetch");
    const res = await server.executeOperation({
      query: Tracks,
      variables: { authorId: "cat-1" },
    });

    expect(fetch).toHaveBeenCalledTimes(2);
    const reqs = fetch.mock.calls.map(([req]) => req);
    expect(new URL(reqs[0].url).search).toContain("author_id=cat-1");
  });
});
