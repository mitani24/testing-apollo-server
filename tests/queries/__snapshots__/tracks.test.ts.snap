// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tracks トラック一覧を取得できる 1`] = `
Object {
  "data": Object {
    "tracks": Array [
      Object {
        "author": Object {
          "id": "string",
          "name": "string",
          "photo": "string",
        },
        "description": "string",
        "id": "string",
        "length": 0,
        "numberOfViews": 0,
        "thumbnail": "string",
        "title": "string",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
